src/common/datomish/transact.cljc

```diff
@@ -552,12 +580,99 @@
     (:db-after (<? (<with db tx-data)))))
 
 (defn <transact!
-  [conn tx-data]
+  "Submits a transaction to the database for writing.
+
+  Returns a pair-chan resolving to `[result error]`."
+  ([conn tx-data]
+   (<transact! conn tx-data (a/chan 1) true))
+  ([conn tx-data result close?]
+   {:pre [(conn? conn)]}
+   ;; Any race to put! is a real race between callers of <transact!.  We can't just park on put!,
+   ;; because the parked putter that is woken is non-deterministic.
+   (let [closed? (not (a/put! (:transact-chan conn) [:sentinel-transact tx-data result close?]))]
+     (go-pair
+       ;; We want to return a pair-chan, no matter what kind of channel result is.
+       (if closed?
+         (raise "Connection is closed" {:error :transact/connection-closed})
+         (<? result))))))
+
+(defn- start-transactor [conn]
+  (let [token-chan (a/chan 1)]
+    (go
+      (>! token-chan (gensym "transactor-token"))
+      (loop []
+        (when-let [token (<! token-chan)]
+          (when-let [[sentinel tx-data result close?] (<! (:transact-chan conn))]
```

Comment:
The big advantage of treating all operations the same, and letting the operation handle closed, is that a new operation (perhaps we want to run queries through this?) is parallel to the existing work.